### PR TITLE
Implement C Memory1 package

### PR DESCRIPTION
**Pull request by kraiouchkine**
## Description

This pull request implements the Memory1 package. `RULE-9-5` is blocked at the moment and not in this PR.

## Change request type

 - [ ] Release or process automation (GitHub workflows, internal scripts)
 - [ ] Internal documentation
 - [ ] External documentation
 - [x] Query files (`.ql`, `.qll`, `.qls` or unit tests)
 - [ ] External scripts (analysis report or other code shipped as part of a release)

## Rules with added or modified queries

 - [ ] No rules added
 - [x] Queries have been added for the following rules:
     - RULE-9-2 - closes github/coding-standards-team#755
     - RULE-9-3 - closes github/coding-standards-team#756
     - RULE-9-4 - closes github/coding-standards-team#757
 - [ ] Queries have been modified for the following rules:
     - _rule number here_

## Release change checklist

A change note ([development_handbook.md#change-notes](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#change-notes)) is required for any pull request which modifies:

 - The structure or layout of the release artifacts.
 - The evaluation performance (memory, execution time) of an existing query.
 - The results of an existing query in any circumstance.

If you are only adding new rule queries, a change note is not required.

_**Author:**_ Is a change note required? 
  - [ ] Yes
  - [x] No

🚨🚨🚨
_**Reviewer:**_ Confirm that format of *shared* queries (not the .qll file, the
.ql file that imports it) is valid by running them within VS Code. 
  - [ ] Confirmed

_**Reviewer:**_ Confirm that either a change note is not required or the change note is required and has been added.
  - [ ] Confirmed

## Query development review checklist

For PRs that add new queries or modify existing queries, the following checklist should be completed by both the author and reviewer:

### Author

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### Reviewer 

 - [x] Have all the relevant rule package description files been checked in?
 - [x] Have you verified that the metadata properties of each new query is set appropriately?
 - [x] Do all the unit tests contain both "COMPLIANT" and "NON_COMPLIANT" cases?
 - [x] Are the alert messages properly formatted and consistent with the [style guide](https://github.com/github/codeql-coding-standards/blob/main/docs/development_handbook.md#query-style-guide)?
 - [x] Have you run the queries on OpenPilot and verified that the performance and results are acceptable?<br />_As a rule of thumb, predicates specific to the query should take no more than 1 minute, and for simple queries be under 10 seconds. If this is not the case, this should be highlighted and agreed in the code review process._
 - [x] Does the query have an appropriate level of in-query comments/documentation?
 - [x] Have you considered/identified possible edge cases?
 - [x] Does the query not reinvent features in the standard library?
 - [x] Can the query be simplified further (not golfed!)

### CODE REVIEW

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

File: rule_packages/c/Memory1.json [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      "properties": {
+        "obligation": "required"
+      },
+      "queries": [
+        {
+          "description": "Using braces in initializers of objects and subobjects improves code readability and clarifies intent.",
+          "kind": "problem",
+          "name": "The initializer for an aggregate or union shall be enclosed in braces",
+          "precision": "high",
+          "severity": "warning",
```
Comment (knewbury01) [inline rule_packages/c/Memory1.json:13]: maybe also have this as recommendation since the cpp rule is

**Review by knewbury01 — COMMENTED** [review]
hey, have only reviewed ~~RULE-9-4~~ (edit) 9-2! so far, so just the one tiny metadata comment, but otherwise looks good, completely agree import,

will finish my review of the other 2 rules on Mon!

File: c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+/**
+ * Holds if `expr1` and `expr2` both initialize the same field of `root`.
+ *
+ * The dbschema keyset for `aggregate_field_init` prevents referencing multiple `Expr`
+ * that initialize the same Field and are part of the same `ClassAggregateLiteral`.
+ * This predicate is therefore unable to distinguish the individual duplicate expressions.
+ */
+predicate hasMultipleInitializerExprsForSameField(ClassAggregateLiteral root, Expr expr, Field f) {
```
Comment (knewbury01) [inline c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql:98]: I think I dont quite understand why this then needs the `Expr` parameter, or the `expr = root.getFieldExpr(f) and` check, can it just say - there is some field initialized more than one time?

```
predicate hasMultipleInitializerExprsForSameField(ClassAggregateLiteral root, Field f) {
  count(root.getFieldExpr(f)) > 1
}
```

**Review by knewbury01 — COMMENTED** [review]
just a few minor comments for RULE-9-4 :) 

will submit a separate review for RULE-9-3!

File: c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql (hunk 2 above)
Comment (kraiouchkine) [reply c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql]: Yes, that was left over from trying to identify the individual expressions. Changes made.

File: c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+ */
+Expr getNthParent(Expr e, int n) {
+  if n = 0 then result = e else result = getNthParent(e.getParent(), n - 1)
+}
+
+/**
+ * Returns a string representation of the index of `e` relative
+ * to the nested array aggregate literal structure it is contained in.
+ */
+string getNestedArrayIndexString(Expr e) {
```
[the remark this answers is not in the document]
Comment (kraiouchkine) [reply c/misra/src/rules/RULE-9-4/RepeatedInitializationOfAggregateObjectElement.ql:32]: Good catch! I forgot to define an explicit array nesting depth order with `order by`. Changes made, and I've updated that test case to cover your example/possible ordering issues.

File: rule_packages/c/Memory1.json (hunk 1 above)
Comment (kraiouchkine) [reply rule_packages/c/Memory1.json]: Change made.

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed but I didn't find anything to test!

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

**Review by knewbury01 — COMMENTED** [review]
heyo, just one small decision on RULE-9-3, I personally am fine with it going either way, but I think I had some like this before and I forget what I was told to do 😅 

other than that, will be ready to approve after that small thing! :) looks good, really great testcase coverage

File: c/misra/src/rules/RULE-9-3/PartiallyInitializedArrayWithExplicitInitializers.ql [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ */
+predicate isMissingExplicitInitializers(AggregateLiteral al) {
+  not al.isCompilerGenerated() and
+  not al.isAffectedByMacro() and
+  // Partially initialized, but not initialized with a leading zero (which is permitted)
+  isPartiallyValueInitialized(al) and
+  not isLeadingZeroInitialized(al)
+}
+
+// note: this query is similar to M8-5-2: MissingExplicitInitializers.ql
```
[the remark this answers is not in the document]
Comment (knewbury01) [reply c/misra/src/rules/RULE-9-3/PartiallyInitializedArrayWithExplicitInitializers.ql:31]: discussed off the thread with @kraiouchkine , I had forgotten that Class includes structs, and the rule explicitly talks about arrays only.

**Review by knewbury01 — APPROVED** [review]
looks good! nice handling of some less than straightforward cases especially with the multidimensional arrays!

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/cpp/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

COMPILE_PASS         : True
RULE                 : M8-5-2
COMPILE_ERROR_OUTPUT : 
TEST_PASS            : True
QUERY                : UseInitBracesToMatchTypeStructure
PACKAGE              : Initialization
SUITE                : AUTOSAR
TEST_DIFFERENCE      : 

COMPILE_PASS         : True
RULE                 : M8-5-2
COMPILE_ERROR_OUTPUT : 
TEST_PASS            : True
QUERY                : NestedZeroValueInitialization
PACKAGE              : Initialization
SUITE                : AUTOSAR
TEST_DIFFERENCE      : 

COMPILE_PASS         : True
RULE                 : M8-5-2
COMPILE_ERROR_OUTPUT : 
TEST_PASS            : True
QUERY                : MissingExplicitInitializers
PACKAGE              : Initialization
SUITE                : AUTOSAR
TEST_DIFFERENCE      : 


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/c/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

SUITE                : MISRA-C-2012
QUERY                : InitializerForAggregateOrUnionNotEnclosedInBraces
COMPILE_ERROR_OUTPUT : 
TEST_PASS            : True
COMPILE_PASS         : True
PACKAGE              : Memory1
TEST_DIFFERENCE      : 
RULE                 : RULE-9-2

SUITE                : MISRA-C-2012
QUERY                : PartiallyInitializedArrayWithExplicitInitializers
COMPILE_ERROR_OUTPUT : 
TEST_PASS            : True
COMPILE_PASS         : True
PACKAGE              : Memory1
TEST_DIFFERENCE      : 
RULE                 : RULE-9-3

SUITE                : MISRA-C-2012
QUERY                : RepeatedInitializationOfAggregateObjectElement
COMPILE_ERROR_OUTPUT : 
TEST_PASS            : True
COMPILE_PASS         : True
PACKAGE              : Memory1
TEST_DIFFERENCE      : 
RULE                 : RULE-9-4


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **clang/c/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

RULE                 : RULE-9-2
TEST_PASS            : True
TEST_DIFFERENCE      : 
QUERY                : InitializerForAggregateOrUnionNotEnclosedInBraces
PACKAGE              : Memory1
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
SUITE                : MISRA-C-2012

RULE                 : RULE-9-3
TEST_PASS            : True
TEST_DIFFERENCE      : 
QUERY                : PartiallyInitializedArrayWithExplicitInitializers
PACKAGE              : Memory1
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
SUITE                : MISRA-C-2012

RULE                 : RULE-9-4
TEST_PASS            : True
TEST_DIFFERENCE      : 
QUERY                : RepeatedInitializationOfAggregateObjectElement
PACKAGE              : Memory1
COMPILE_ERROR_OUTPUT : 
COMPILE_PASS         : True
SUITE                : MISRA-C-2012


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! **gcc/cpp/x86_64** Matrix Testing for this PR has been completed. See below for the results! <br><br>
``` 

TEST_PASS            : True
QUERY                : UseInitBracesToMatchTypeStructure
RULE                 : M8-5-2
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
COMPILE_ERROR_OUTPUT : 
SUITE                : AUTOSAR
PACKAGE              : Initialization

TEST_PASS            : True
QUERY                : MissingExplicitInitializers
RULE                 : M8-5-2
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
COMPILE_ERROR_OUTPUT : 
SUITE                : AUTOSAR
PACKAGE              : Initialization

TEST_PASS            : True
QUERY                : NestedZeroValueInitialization
RULE                 : M8-5-2
TEST_DIFFERENCE      : 
COMPILE_PASS         : True
COMPILE_ERROR_OUTPUT : 
SUITE                : AUTOSAR
PACKAGE              : Initialization


```

**Comment by jsinglet** [comment]
🤖 Beep Boop! Matrix Testing for this PR has been **completed**. If no reports were posted it means this PR does not contain things that need matrix testing!

**Pull request merged**